### PR TITLE
[EH] Use _UA_SEARCH_PHASE in personality function (NFC)

### PR DESCRIPTION
**Pull request by aheejin**
In two-phase unwinding, the first phase is the search phase (`_UA_SEARCH_PHASE`) and the second one is the cleanup phase (`_UA_CLEANUP_PHASE`). The search phase searches up the stack to see if there is a matching catch handler, and if it finds one, it caches the result. And in the second cleanup phase, it retrieves the cached result (to avoid doing same work twice) and unwinds the stack.

Wasm does not do the two-phase unwinding; it only has a single phase. We used `_UA_CLEANUP_PHASE` for this single phase, so in Wasm the cleanup phase is supposed to the search. So we several many custom `#ifdef`s to use the code guarded by `_UA_SEARCH_PHASE`, for example: https://github.com/aheejin/emscripten/blob/d57db5bea1719319a680699c50b91fa3d88fa0ec/system/lib/libcxxabi/src/cxa_personality.cpp#L771-L776 https://github.com/aheejin/emscripten/blob/d57db5bea1719319a680699c50b91fa3d88fa0ec/system/lib/libcxxabi/src/cxa_personality.cpp#L850-L855

These are apparently gone in #14288, which replaced many `if`s with `assert`s. This in effect removed our special handling for `_UA_CLEANUP_PHASE`; there are several `assert`s that asserts the current phase is `_UA_SEARCH_PHASE`, while Wasm is in `_UA_CLEANUP_PHASE`. But this has not caused problems so far because we have built libc++abi with `-NDEBUG`, so all assertions were no-op. https://github.com/emscripten-core/emscripten/blob/40fb7d2071e439f1de614898b88518df582faa94/tools/system_libs.py#L1366 But this is now a problem because #17979 adds a debug build of libc++abi, which enables assertions.

Come to think of it, I'm not sure why I decided to use `_UA_CLEANUP_PHASE` for our single phase in the first place. If we use `_UA_SEARCH_PHASE`, we can remove more our custom code and reduce the difference between our port and the upstream library.

### CODE REVIEW

**Review by sbc100 — APPROVED** [review]
I don't fully understand the code but less code sounds good to me!

File: system/lib/libcxxabi/src/cxa_personality.cpp [hunk 1]
```diff
@@ -964,6 +966,11 @@ __gxx_personality_v0
             exc->languageSpecificData = results.languageSpecificData;
             exc->catchTemp = reinterpret_cast<void*>(results.landingPad);
             exc->adjustedPtr = results.adjustedPtr;
+#ifdef __USING_WASM_EXCEPTIONS__
+            // Wasm only uses a single phase (_UA_SEARCH_PHASE), so save the
+            // results here.
+            set_registers(unwind_exception, context, results);
+#endif
```
Comment (aheejin) [inline system/lib/libcxxabi/src/cxa_personality.cpp:973]: This is to save our selector computation result.

In the upstream version, the search phase searches for a matching catch handler and caches the result, and the second cleanup phase retrieves the result and sets registers with the final selector value. (In Wasm's libunwind, they are not registers but just a field in a struct.). They do it in the second cleanup phase: https://github.com/emscripten-core/emscripten/blob/40fb7d2071e439f1de614898b88518df582faa94/system/lib/libcxxabi/src/cxa_personality.cpp#L973

But now we only use the search phase, we do it here. Instead, we are able to remove our duplicated caching code in line 984-993 below.

File: system/lib/libcxxabi/src/cxa_personality.cpp [hunk 2]
```diff
@@ -981,16 +988,6 @@ __gxx_personality_v0
...
-#ifdef __USING_WASM_EXCEPTIONS__
-    // Wasm uses only one phase in _UA_CLEANUP_PHASE, so we should set
-    // these here.
-    __cxa_exception* exception_header = (__cxa_exception*)(unwind_exception+1) - 1;
-    exception_header->handlerSwitchValue = static_cast<int>(results.ttypeIndex);
-    exception_header->actionRecord = results.actionRecord;
-    exception_header->languageSpecificData = results.languageSpecificData;
-    exception_header->catchTemp = reinterpret_cast<void*>(results.landingPad);
-    exception_header->adjustedPtr = results.adjustedPtr;
-#endif
```
Comment (aheejin) [inline system/lib/libcxxabi/src/cxa_personality.cpp:993]: This is what the upstream code does in the first search phase, but we used only the cleanup phase, we duplicated the code here. Now we can remove it.

File: system/lib/libcxxabi/src/cxa_personality.cpp [hunk 3]
```diff
@@ -798,6 +798,8 @@ static void scan_eh_tab(scan_results &results, _Unwind_Action actions,
                         {
                             // Native exception caught by exception
                             // specification.
+                            assert(actions & _UA_SEARCH_PHASE);
+                            results.ttypeIndex = ttypeIndex;
```
Comment (aheejin) [inline system/lib/libcxxabi/src/cxa_personality.cpp:802]: This is not related to this PR, but I compared our `cxa_personality.cpp` with llvm 14.0.0 version, this part is missing. I guess it was dropped when we upgraded the library. I don't think Wasm is using this part of the code though, reducing difference from the upstream code seems better.

**Pull request merged**